### PR TITLE
Use Zookeeper persistent recursive Zookeeper watches

### PR DESCRIPTION
**Pull request by noblepaul**
In newer versions of Zookeeper there is a concept of persistent watches. This could provide a lot of value to Solr and PRS state specifically given the ability to do a recursive watch on a parent node and get updates for child nodes. This also enables us to update only a single PRS node instead of reading all at once

https://zookeeper.apache.org/doc/current/zookeeperProgrammers.html#sc_WatchPersistentRecursive

### CODE REVIEW

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 1]
```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
...
+                  log.trace("fresh replica, force fetch all {}", collectionPath);
+                }
+
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
+              } else {
+                // the PRS entry is already available, need to do an update
+                Stat stat = null;
+                try {
+                  // get the stat of this child node
+                  stat = zkClient.exists(path, null, true);
```
Comment (hiteshk25) [inline solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:452]: What if we get stat of `state.json` here?

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 2]
```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
...
+                }
+                // removed the duplicate without modifying the PRS Object
+                oldState.removeDuplicate(newState);
+                return;
+              } else {
+                if (log.isTraceEnabled()) {
+                  log.trace("Replica:  {} delete, force fetch", path);
+                }
+                // a replica got removed. we can't get the pzxid reliably. fetch everything
+                prs = PerReplicaStatesFetcher.fetch(collectionPath, zkClient, null);
```
Comment (hiteshk25) [inline solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:487]: similar, can we get stat of `state.json` and use cversion of it?

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java (hunk 1 above)
Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:452]: that's incorrect . The stat of `state.json` would've gotten changed by another event

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java (hunk 2 above)
Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:487]: `cversion` is not used anymore at all. it's unreliable

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 3]
```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
...
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.error("exception for collection: [{}]", coll, e);
+        }
+        collectionWatches.updateDocCollection(coll, newState);
+        synchronized (getUpdateLock()) {
+          constructState(Collections.singleton(coll));
+        }
+      }
+      if (collectionState != null && collectionState.isPerReplicaState()) {
```
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:417]: yeah. we could do a final refactoring once everything else is right

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 4]
```diff
@@ -326,14 +326,14 @@ private boolean updateDocCollection(String collection, DocCollection newState) {
               log.debug("Removing cached collection state for [{}]", collection);
               watch.currentState = null;
             } else { // both new and old states are non-null
-              int oldCVersion =
+              long oldCVersion =
```
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:329]: yes, we should

File: solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java [hunk 5]
```diff
@@ -368,10 +368,152 @@ private StatefulCollectionWatch compute(
...
+          PerReplicaStates prs = collectionState.getPerReplicaStates();
+          if (prs != null) {
+            String stateStr = path.substring(collectionPath.length() + 1);
+            PerReplicaStates.State newState = PerReplicaStates.State.parse(stateStr);
+            PerReplicaStates.State oldState = prs.states.get(newState.replica);
+            if (event.getType() == EventType.NodeCreated) {
+              if (oldState != null && newState.version < oldState.version) {
+                // we got a notification out of order? . Shouldn't happen
+                if (log.isTraceEnabled()) {
+                  log.trace("newState {} < oldState {}", newState, oldState);
```
[the remark this answers is not in the document]
Comment (noblepaul) [reply solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java:435]: yes. we should

**Pull request closed without merging**